### PR TITLE
Haveibeenpwnd review

### PR DESCRIPTION
**Pull request by dfish3r**

### CODE REVIEW

File: src/main/java/org/passay/HaveIBeenPwnedRule.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+import java.time.Duration;
+import java.util.Collections;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.cryptacular.codec.HexEncoder;
+import org.cryptacular.util.CodecUtil;
+import org.cryptacular.util.HashUtil;
+
+/**
+ * Validates if the password against the online database of <code>haveibeenpwned.com</code>
```
Comment (serac) [inline src/main/java/org/passay/HaveIBeenPwnedRule.java:21]:
```suggestion
 * Validates the password against the online database of <code>haveibeenpwned.com</code>
```

File: src/main/java/org/passay/HaveIBeenPwnedRule.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+   *
+   * @return reader for the connection
+   *
+   * @throws IOException if a connection cannot be opened to the API
+   */
+  private LineNumberReader openApiConnectionForRange(final String range)
+    throws IOException
+  {
+    final URL url = new URL(apiUrl, range);
+    final URLConnection c = url.openConnection();
```
Comment (serac) [inline src/main/java/org/passay/HaveIBeenPwnedRule.java:216]: May want to consider HTTP connection pooling here. That said I appreciate the simplicity of this code using JDK libraries.

File: src/main/java/org/passay/HaveIBeenPwnedRule.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+   * @param reader to read
+   *
+   * @return rule result whose validity is determined by whether a match was found
+   *
+   * @throws IOException if an error occurs reading from the reader
+   */
+  private RuleResult searchResponse(final String hexDigest, final LineNumberReader reader) throws IOException
+  {
+    String line;
+    final Pattern p = Pattern.compile("^(" + hexDigest.substring(PREFIX_LENGTH) + "):(\\d+)\\s*$");
```
Comment (serac) [inline src/main/java/org/passay/HaveIBeenPwnedRule.java:175]: I don't believe substring matching is correct here. According to https://haveibeenpwned.com/API/v2#PwnedPasswords:

> When a password hash with the same first 5 characters is found in the Pwned Passwords repository, the API will respond with an HTTP 200 and include the suffix of every hash beginning with the specified prefix, followed by a count of how many times it appears in the data set. The API consumer can then search the results of the response for the presence of their source hash and if not found, the password does not exist in the data set.

I take that to mean the _entire_ source has should be matched via simple startsWith semantics.

**Review by serac — CHANGES_REQUESTED** [review]
I believe the `searchResponse` method is liable to false positives and cited the API docs about best practices for matching.

File: src/main/java/org/passay/HaveIBeenPwnedRule.java (hunk 2 above)
Comment (dfish3r) [reply src/main/java/org/passay/HaveIBeenPwnedRule.java:216]: Seems like a v2 feature. Note that persistent connections are supported:
https://docs.oracle.com/javase/8/docs/technotes/guides/net/http-keepalive.html

File: src/main/java/org/passay/HaveIBeenPwnedRule.java (hunk 3 above)
Comment (serac) [reply src/main/java/org/passay/HaveIBeenPwnedRule.java:175]: I misunderstood both your code and the data returned. I read your code as only examining the _start_ of data, but I see it actually matches the _suffix_ of data, which is what is returned by the API. Thus my concerns were off base.

**Pull request merged**